### PR TITLE
Moe support

### PR DESCRIPTION
**Pull request by cehongwang**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant and/or add your own.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/select.py
Comment (peri044) [inline py/torch_tensorrt/dynamo/conversion/impl/select.py:1]: Could you add detailed comments demonstrating what it does now that you've gone through the entire converter ? that would be helpful

File: tools/llm/run_llm.py [hunk 1]
```diff
@@ -71,7 +71,7 @@ def get_model(args):
     else:
         model = model.to(torch.float32)
 
-    return model
+    return model.cuda()
```
Comment (peri044) [inline tools/llm/run_llm.py]: model.cuda() is already done during initialization

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 2]
```diff
@@ -592,13 +591,11 @@ def _save_weight_mapping(self) -> None:
         torch.cuda.empty_cache()
 
     @needs_refit  # type: ignore[misc]
-    def _insert_engine_to_cache(self, hash_val: str, serialized_engine: bytes) -> None:
+    def _insert_engine_to_cache(self, hash_val: str, engine: bytes) -> None:
+        serialized_engine = engine.serialize()
         # TODO: @Evan is waiting for TRT's feature to cache the weight-stripped engine
         # if not self.compilation_settings.strip_engine_weights:
         #     # set EXCLUDE_WEIGHTS flag to strip weights
-        #     runtime = trt.Runtime(TRT_LOGGER)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:599]: why are these being deleted?

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 3]
```diff
@@ -182,7 +183,19 @@ def __init__(
...
+        if cuda_engine:
+            assert isinstance(
+                cuda_engine, trt.ICudaEngine
+            ), "Cuda engine must be a trt.ICudaEngine object"
+            self.engine = cuda_engine
+        elif serialized_engine:
+            assert isinstance(
+                serialized_engine, bytes
+            ), "Serialized engine must be a bytes object"
+            self.engine = serialized_engine
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: deserialize the engine here

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py [hunk 4]
```diff
@@ -772,14 +768,10 @@ def run(
...
-            self._insert_engine_to_cache(hash_val, serialized_engine)
-
-        with io.BytesIO() as engine_bytes:
-            engine_bytes.write(serialized_engine)
-            engine_str = engine_bytes.getvalue()
+            self._insert_engine_to_cache(hash_val, cuda_engine)
 
         return TRTInterpreterResult(
-            engine_str,
+            cuda_engine,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py]: How will this work with the deferred engine setup?

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py:1]: Generally I think this is not clearly implemented. If we decided that PyRuntime will use initialized engines, then you should quickly reduce other cases to an initialized engine.

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py (hunk 3 above)
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: Also do not overload the meaning of `self.engine`. Either it should be the serialized engine or the live engine not both

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 5]
```diff
@@ -260,13 +272,22 @@ def set_default_device_memory_budget(self) -> int:
         return self._set_device_memory_budget(budget_bytes)
 
     def setup_engine(self) -> None:
+
+        if isinstance(self.engine, trt.ICudaEngine):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: We need to think about how this will interact with the greater system like the `lazy_engine_init` option.

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 6]
```diff
@@ -260,13 +272,22 @@ def set_default_device_memory_budget(self) -> int:
...
+            pass
+        elif isinstance(self.engine, bytes):
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: we should not check by type, have two members and a state flag

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 7]
```diff
@@ -302,7 +323,7 @@ def _check_initialized(self) -> None:
             raise RuntimeError("PythonTorchTensorRTModule is not initialized.")
 
     def _on_state_dict(self, state_dict: Dict[str, Any], prefix: str, _: Any) -> None:
-        state_dict[prefix + "engine"] = self.serialized_engine
+        state_dict[prefix + "engine"] = self.engine
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: Same thing here, this should just be serialized engine

File: py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py [hunk 8]
```diff
@@ -317,7 +338,7 @@ def _load_from_state_dict(
         unexpected_keys: Any,
         error_msgs: Any,
     ) -> None:
-        self.serialized_engine = state_dict[prefix + "engine"]
+        self.engine = state_dict[prefix + "engine"]
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_PythonTorchTensorRTModule.py]: ^^

File: py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py [hunk 9]
```diff
@@ -76,6 +78,7 @@ class TorchTensorRTModule(torch.nn.Module):  # type: ignore[misc]
 
     def __init__(
         self,
+        cuda_engine: Optional[trt.ICudaEngine | bytes] = None,
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/runtime/_TorchTensorRTModule.py]: Should not be a union. Its either you give a cuda_engine or a serialized engine

File: py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py (hunk 2 above)
Comment (cehongwang) [reply py/torch_tensorrt/dynamo/conversion/_TRTInterpreter.py:599]: Don't need to deserialize the engine anymore. The engine is already live

File: py/torch_tensorrt/dynamo/lowering/passes/remove_num_users_is_0_nodes.py [hunk 10]
```diff
@@ -23,7 +23,8 @@ def remove_num_users_is_0_nodes(
             and len(node.all_input_nodes) > 0
         ):
             gm.graph.erase_node(node)
-            gm = clean_up_graph_after_modifications(gm)
+
+    gm = clean_up_graph_after_modifications(gm)
```
Comment (narendasan) [inline py/torch_tensorrt/dynamo/lowering/passes/remove_num_users_is_0_nodes.py:27]: Is this just because we only need one cleanup?

Comment (cehongwang) [reply py/torch_tensorrt/dynamo/lowering/passes/remove_num_users_is_0_nodes.py:27]: Yeah or otherwise it could take forever

File: tests/py/dynamo/conversion/test_index_put_aten.py [hunk 11]
```diff
@@ -194,11 +195,43 @@ class TestIndexPutConverter(DispatchTestCase):
                     dtype=torch.int32,
                 ),
             ),
+            # param(
+            #     test_name="4d_indices_none_none_multiple_idx_broadcast_error",
+            #     source_tensor=torch.zeros([1, 2, 5, 3], dtype=torch.float32),
+            #     indices_tensor=(None, None, torch.tensor([0, 1, 2], dtype=torch.int64)),
+            #     value_tensor=torch.randn([2, 3, 3], dtype=torch.float32),
+            # ),
```
Comment (peri044) [inline tests/py/dynamo/conversion/test_index_put_aten.py:203]: Consider mentioning the limitations of this converter and why this case is not supported ?

File: tests/py/dynamo/conversion/test_index_put_aten.py [hunk 12]
```diff
@@ -244,6 +277,94 @@ def forward(self, source_tensor, value_tensor):
             use_dynamo_tracer=True,
         )
 
+    def test_index_add_dynamic_shape(self):
+
+        class Model(torch.nn.Module):
+            def forward(self, x, y, z, a, b):
+                x.index_add_(0, y, z)
+                x.index_add_(0, a, b)
+                return x
```
Comment (peri044) [inline tests/py/dynamo/conversion/test_index_put_aten.py:286]: Does this testcase produce index_put ops ? or how is this connected ?

File: tests/py/dynamo/conversion/test_index_put_aten.py (hunk 11 above)
Comment (cehongwang) [reply tests/py/dynamo/conversion/test_index_put_aten.py:203]: idk who wrote this test case but pytorch gives an error for this test case

File: tests/py/dynamo/conversion/test_index_put_aten.py (hunk 12 above)
Comment (cehongwang) [reply tests/py/dynamo/conversion/test_index_put_aten.py:286]: Yes. index_add_ uses aten.index_put

File: tests/py/dynamo/conversion/test_index_put_aten.py (hunk 11 above)
Comment (narendasan) [reply tests/py/dynamo/conversion/test_index_put_aten.py:203]: Seems like @chohk88

**Comment by peri044** [comment]
@cehongwang the linux tests are failing. Do you know the rootcause ?

**Comment by cehongwang** [comment]
> @cehongwang the linux tests are failing. Do you know the rootcause ?

fixed
